crates/datasources/src/object_store/glob_util.rs

```diff
@@ -0,0 +1,155 @@
+            .ok_or_else(|| ReplacerFindError::new(format!("expected range N..M: {pattern}")))?;
+
+        let start: i64 = start.parse().map_err(|e| {
+            ReplacerFindError::new(format!("start in range not a valid number: {e}"))
+        })?;
+        let end: i64 = end
+            .parse()
+            .map_err(|e| ReplacerFindError::new(format!("end in range not a valid number: {e}")))?;
+
+        (start..=end).map(|i| i.to_string()).collect::<Vec<_>>()
+    } else {
+        vec![validate_string(pattern)?]
+    };
+
+    Ok(Replacer {
+        start,
+        end,
+        iter: ranges.into_iter(),
+    })
+}
+
+pub fn get_resolved_patterns(pattern: String) -> Vec<ResolvedPattern> {
+    let Replacer { start, end, iter } = match get_replacer(&pattern) {
+        Ok(replacer) => replacer,
+        Err(error) => {
+            debug!(%error, %pattern, "cannot resolve pattern");
+            return vec![ResolvedPattern(pattern.to_owned())];
+        }
+    };
+
+    iter.flat_map(|replacement| {
+        let begin = &pattern[0..start];
+        let end = &pattern[end + 1..pattern.len()];
+        let new_pattern = format!("{begin}{replacement}{end}");
+        get_resolved_patterns(new_pattern)
+    })
+    .collect()
+}
+
+#[derive(Debug, Clone)]
+pub struct ResolvedPattern(String);
+
+impl From<ResolvedPattern> for String {
+    fn from(ResolvedPattern(string): ResolvedPattern) -> Self {
+        string
+    }
+}
+
+impl AsRef<str> for ResolvedPattern {
+    fn as_ref(&self) -> &str {
+        let ResolvedPattern(string) = self;
+        string
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::get_resolved_patterns;
+
+    #[test]
+    fn test_resolve_glob_patterns() {
+        let test_cases: &[(&str, &[&str])] = &[
+            ("file_{3..6}", &["file_3", "file_4", "file_5", "file_6"]),
+            ("file_{x,y,z}", &["file_x", "file_y", "file_z"]),
+            (
+                "file_0{1..3}.csv",
+                &["file_01.csv", "file_02.csv", "file_03.csv"],
+            ),
+            (
+                "file_{a-b,c-d,e-f}.json",
+                &["file_a-b.json", "file_c-d.json", "file_e-f.json"],
+            ),
+            (
+                "file_{0..2}_{p,q,r}.parquet",
+                &[
+                    "file_0_p.parquet",
+                    "file_0_q.parquet",
+                    "file_0_r.parquet",
+                    "file_1_p.parquet",
+                    "file_1_q.parquet",
+                    "file_1_r.parquet",
+                    "file_2_p.parquet",
+                    "file_2_q.parquet",
+                    "file_2_r.parquet",
+                ],
+            ),
+        ];
+
+        for (pattern, expected) in test_cases {
+            let patterns = get_resolved_patterns(pattern.to_string());
+            let actual = patterns.iter().map(AsRef::as_ref).collect::<Vec<_>>();
+            assert_eq!(&actual, expected);
+        }
+
+        // TODO: Add test cases for failed patterns...
```

Comment:
Added more cases to handle the cases you mentioned.

I am not very keen on throwing an error since the "pattern" formation can be coincidental and we shouldn't throw an error until user opts-in or opts-out.

We definitely should have an "opt-out" from glob pattern so we can throw errors with confidence.